### PR TITLE
Core: Add story preloading to optimize lazy compilation

### PR DESCRIPTION
**Pull request by ndelangen**
## What I did

When using lazy-compilation, loading a story can take a bit of time, this tries to optimize that experience a bit, by guessing which story you'll visit next, and sending a command to the preview to preload the related stories ahead of time.

### CODE REVIEW

**Review by shilman — COMMENTED** [review]
@ndelangen what's the best way to QA this?

File: lib/api/src/modules/stories.ts [hunk 1]
```diff
@@ -439,12 +446,25 @@ export const init: ModuleFn = ({
...
         if (options) {
           checkDeprecatedOptionParameters(options);
           fullAPI.setOptions(options);
         }
+
+        // create a list of related stories to be preloaded
+        const toBePreloaded = Array.from(
+          new Set([
+            api.findSiblingStoryId(storyId, storiesHash, 1, false),
+            api.findSiblingStoryId(storyId, storiesHash, -1, false),
```
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:461]: There's no need to preload the sibling stories of the same component, as they all come from the same CSF file, and thus will already be loaded. I guess there's no harm in doing so? Not sure.

File: lib/preview-web/src/PreviewWeb.tsx [hunk 2]
```diff
@@ -398,6 +399,10 @@ export class PreviewWeb<TFramework extends AnyFramework> {
     await this.onUpdateArgs({ storyId, updatedArgs });
   }
 
+  async onPreloadStory(ids: string[]) {
```
Comment (tmeasday) [inline lib/preview-web/src/PreviewWeb.tsx:402]: If it's an array of ids, should we call this `preloadStories`?

File: lib/preview-web/src/PreviewWeb.tsx [hunk 3]
```diff
@@ -127,6 +127,7 @@ export class PreviewWeb<TFramework extends AnyFramework> {
     this.channel.on(Events.RESET_STORY_ARGS, this.onResetArgs.bind(this));
     this.channel.on(Events.FORCE_RE_RENDER, this.onForceReRender.bind(this));
     this.channel.on(Events.FORCE_REMOUNT, this.onForceRemount.bind(this));
+    this.channel.on(Events.STORY_PRELOAD, this.onPreloadStory.bind(this));
```
Comment (tmeasday) [inline lib/preview-web/src/PreviewWeb.tsx:130]: I think the naming for these imperative events would imply this should be `PRELOAD_STORIES`

**Review by tmeasday — CHANGES_REQUESTED** [review]
LGTM apart from doing unnecessary work.

Can we have a test in `PreviewWeb` that when it receives the event it loads the stories?

You can do something simple like:

```js
const preview = await createAndRenderPreview();

importFn.mockReset();
await preview.onPreloadStory(['component-two--c'])

expect(importFn).toHaveBeenCalledWith(...);
```

**Comment by ndelangen** [comment]
To test in full, we need an example or repro with lazyCompilation enabled

File: lib/api/src/modules/stories.ts (hunk 1 above)
Comment (ndelangen) [reply lib/api/src/modules/stories.ts:461]: fair, I will change this.

File: lib/preview-web/src/PreviewWeb.tsx (hunk 3 above)
Comment (ndelangen) [reply lib/preview-web/src/PreviewWeb.tsx]: done!

File: lib/preview-web/src/PreviewWeb.tsx (hunk 2 above)
Comment (ndelangen) [reply lib/preview-web/src/PreviewWeb.tsx]: done!

**Comment by ndelangen** [comment]
This can be tested in the react-ts example. @shilman and I tested it together 🎉

File: lib/core-events/src/index.ts [hunk 4]
```diff
@@ -16,6 +16,8 @@ enum events {
   FORCE_RE_RENDER = 'forceReRender',
   // Force the current story to re-render from scratch, with its initial args
   FORCE_REMOUNT = 'forceRemount',
+  // Request the story has been loaded into the store, ahead of time, before it's actually
+  PRELOAD_STORIES = 'storyPreload',
```
Comment (tmeasday) [inline lib/core-events/src/index.ts:20]:
```suggestion
  PRELOAD_STORIES = 'preloadStories',
```

File: lib/preview-web/src/PreviewWeb.test.ts [hunk 5]
```diff
@@ -1022,6 +1022,23 @@ describe('PreviewWeb', () => {
...
+        Promise.resolve({ default: { title: 'Component Two' }, C: () => {} } as any)
+      );
+      const preview = await createAndRenderPreview({
+        importFn: mockedImportFn,
+      });
+      await preview.onPreloadStories(['component-two--c']);
+
+      await waitForRender();
+
+      expect(mockedImportFn).toHaveBeenCalledWith('./src/ComponentTwo.stories.js');
```
Comment (tmeasday) [inline lib/preview-web/src/PreviewWeb.test.ts:1038]: This a little confusing as the initial selection `component-one--a` isn't actually going to be returned from the first render, I guess it shows an error but we aren't checking? 

```suggestion
      document.location.search = '?id=component-one--a&viewMode=docs';
      const preview = await createAndRenderPreview();
      await waitForRender();

      importFn.mockClear();
      await preview.onPreloadStories(['component-two--c']);
      expect(importFn).toHaveBeenCalledWith('./src/ComponentTwo.stories.js');
```

I think this is a bit simpler?

**Review by tmeasday — APPROVED** [review]
Thanks @ndelangen! Couple nitpicks

**Comment by yairEO** [comment]
> by guessing which story you'll visit next

Would have been nice to have included a short explanation of that logic

16 years ago I used to preload web pages when the mouse was hovering a link `<a>` more than N milliseconds threshold, so that's a good way of guessing where the user might click next

**Comment by ndelangen** [comment]
@yairEO I'm preloading stories which would be navigated to if the user used the keyboard shortcuts to jump to the next and previous component.

**Comment by ndelangen** [comment]
Great recommendation on the hover, would you be interested in implementing that as a contribution?

I could do it, but if it's something you'd be keen to do as well, I could assist! LMK

File: lib/core-events/src/index.ts [hunk 6]
```diff
@@ -16,6 +16,8 @@ enum events {
   FORCE_RE_RENDER = 'forceReRender',
   // Force the current story to re-render from scratch, with its initial args
   FORCE_REMOUNT = 'forceRemount',
+  // Request the story has been loaded into the store, ahead of time, before it's actually
```
Comment (yairEO) [inline lib/core-events/src/index.ts:19]: The sentence ends abruptly :) *before it's actually* what?

Also, that is a missing word between `story` and `has`, like `which`

**Comment by yairEO** [comment]
> Great recommendation on the hover, would you be interested in implementing that as a contribution?
> 
> I could do it, but if it's something you'd be keen to do as well, I could assist! LMK

I would prefer you do it because it would take me a long long time to understand the codebase and I also don't know typescript... I try to avoid typescript until it dies like coffescript (hopefully!)

**Pull request merged**